### PR TITLE
Instance segmentation papercut

### LINKED ISSUE #930: resize instance segmentation mask
## 🐛 Bug

found another paper-cut with instance segmentation...
it is not trivial/intuitive how to match mask with input image as output can be any size and mask is 128x128
so it is just scaled equally in each dimension is applied padding uniformly?

### To Reproduce

https://www.kaggle.com/jirkaborovec/cell-instance-segm-lightning-flash#Training-with-Flash-Lightning

### PR DESCRIPTION
**Pull request by gianscarpe**
## What does this PR do?

Fixes #930

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/integrations/icevision/transforms.py [hunk 1]
```diff
@@ -239,9 +240,12 @@ def __init__(self, transform: List[Callable]):
         self.transform = A.Adapter(transform)
 
     def forward(self, x):
+
         record = to_icevision_record(x)
         record = self.transform(record)
-        return from_icevision_record(record)
+        record = from_icevision_record(record)
+        record[DataKeys.METADATA]["original_size"] = x[DataKeys.INPUT].shape[:-1]
```
Comment (ethanwharris) [inline flash/core/integrations/icevision/transforms.py:247]: This should probably live within the input object rather than in the transform adapter as the adapter may be used multiple times

File: flash/image/instance_segmentation/data.py [hunk 2]
```diff
@@ -12,23 +12,125 @@
...
+class InstanceSegmentationNumpyInput(NumpyInput):
+    def load_sample(self, sample: Dict[str, Any], dataset: Optional[Any] = None) -> Dict[str, Any]:
+        breakpoint()
+        img = torch.from_numpy(sample[DataKeys.INPUT]).float()
+        sample[DataKeys.INPUT] = img
+        sample[DataKeys.METADATA] = {"size": img.shape}
+        return sample
+
+
+class InstanceSegmentationPathsInput(PathsInput):
```
Comment (ethanwharris) [inline flash/image/instance_segmentation/data.py:53]: I don't think this is needed (and it doesn't seem to be used).

File: flash/image/instance_segmentation/data.py [hunk 3]
```diff
@@ -12,23 +12,125 @@
...
     COCOMaskParser = object
     VOCMaskParser = object
     Parser = object
+if _KORNIA_AVAILABLE:
+    import kornia as K
+
+
+class InstanceSegmentationNumpyInput(NumpyInput):
+    def load_sample(self, sample: Dict[str, Any], dataset: Optional[Any] = None) -> Dict[str, Any]:
+        breakpoint()
```
Comment (ethanwharris) [inline flash/image/instance_segmentation/data.py:46]:
```suggestion
```

File: flash/image/instance_segmentation/data.py [hunk 4]
```diff
@@ -12,23 +12,125 @@
...
+
 else:
     COCOMaskParser = object
     VOCMaskParser = object
     Parser = object
+if _KORNIA_AVAILABLE:
+    import kornia as K
+
+
+class InstanceSegmentationNumpyInput(NumpyInput):
```
Comment (ethanwharris) [inline flash/image/instance_segmentation/data.py:44]: I think the segmentation numpy input now looks a bit different on master. Not sure how this will play with icevision, I think it will work for prediction only which is fine if that is the intended use case :smiley:

File: flash/image/instance_segmentation/data.py [hunk 5]
```diff
@@ -74,15 +177,24 @@ def train_default_transforms(self) -> Optional[Dict[str, Callable]]:
...
+    # def uncollate(batch: Any) -> Any:
+    #     breakpoint()
+    #     return batch[DataKeys.PREDS]
+    def per_sample_transform(self, sample: Any) -> Any:
+        # TODO GET HERE WITH image SIZE
+        breakpoint()
+        resize = K.geometry.Resize(sample[DataKeys.METADATA]["original_size"][-2:], interpolation="bilinear")
+        sample[DataKeys.PREDS]["mask_array"] = (
+            resize(torch.tensor(sample[DataKeys.PREDS]["mask_array"], dtype=torch.float)) > 0
+        ).to(torch.uint8)
```
Comment (ethanwharris) [inline flash/image/instance_segmentation/data.py:190]: The mask can have multiple classes so the `> 0` won't have the desired effect here as it casts everything to binary. I think a better way to do it would be with nearest. neighbour interpolation then just cast to a long tensor.

File: flash/image/instance_segmentation/model.py [hunk 6]
```diff
@@ -19,6 +19,7 @@
 from flash.core.data.data_pipeline import DataPipeline
 from flash.core.data.output import PredsOutput
 from flash.core.registry import FlashRegistry
+from flash.core.utilities.imports import _KORNIA_AVAILABLE
```
Comment (ethanwharris) [inline flash/image/instance_segmentation/model.py:22]: Looks unused:
```suggestion
```

File: flash/image/instance_segmentation/data.py [hunk 7]
```diff
@@ -74,15 +177,24 @@ def train_default_transforms(self) -> Optional[Dict[str, Callable]]:
...
+        return super().per_sample_transform(sample)
 
 
 class InstanceSegmentationData(DataModule):
 
     input_transform_cls = InstanceSegmentationInputTransform
-    output_transform_cls = InstanceSegmentationOutputTransform
 
     @classmethod
     def from_folders(
```
Comment (ethanwharris) [inline flash/image/instance_segmentation/data.py:200]: Let's also add a `from_numpy` to the datamodule which only has a `predict_array` argument (not train, test, val) :smiley: Actually, should we add this to all icevision tasks???

**Review by ethanwharris — COMMENTED** [review]
Looking good! Predict from numpy will be a really cool addition :smiley:
